### PR TITLE
replace js annotation typeclass method

### PR DESCRIPTION
**Pull request by chikeabuah**

### CODE REVIEW

File: hs/src/Reach/JSUtil.hs [hunk 1]
```diff
@@ -231,6 +265,46 @@ instance HasJSAnnot JSBinOp where
...
+    JSArrayElement e -> jsa e
+    JSArrayComma a -> a
+
+instance RepJSAnnot JSArrayElement where
+  rjsa a' = \case
+    JSArrayElement e -> JSArrayElement $ rjsa a' e
+    JSArrayComma _ -> JSArrayComma a'
+
+instance RepJSAnnot JSObjectPropertyList where
+  rjsa a' x = (mkCommaTrailingList (map (rjsa a') $ jso_flatten x))
```
Comment (jeapostrophe) [inline hs/src/Reach/JSUtil.hs:306]: I don't think you should be flattening and making... you should just traverse it here

File: hs/src/Reach/JSUtil.hs [hunk 2]
```diff
@@ -177,6 +178,12 @@ a2sp = JSSemi
 class HasJSAnnot a where
   jsa :: a -> JSAnnot
 
+class RepJSAnnot a where
+  rjsa :: JSAnnot -> a -> a
+
+rjsaJSL :: (HasJSAnnot a, RepJSAnnot a) => JSAnnot -> JSCommaList a -> JSCommaList a
+rjsaJSL a' y =  toJSCL $ map (rjsa a') $ jscl_flatten y
```
Comment (jeapostrophe) [inline hs/src/Reach/JSUtil.hs:185]: I think with `RepJSAnnot`, you should just implement `class RepJSAnnot a => RepJSAnnot (JSCommaList a)`

File: hs/src/Reach/JSUtil.hs [hunk 3]
```diff
@@ -268,19 +342,69 @@ instance HasJSAnnot JSExpression where
...
+instance RepJSAnnot JSExpression where
+  rjsa a' = \case
+    JSIdentifier _ x -> JSIdentifier a' x
+    JSDecimal _ x -> JSDecimal a' x
+    JSLiteral _ x -> JSLiteral a' x
+    JSHexInteger _ x -> JSHexInteger a' x
+    JSOctal _ x -> JSOctal a' x
+    JSStringLiteral _ x -> JSStringLiteral a' x
+    JSRegEx _ x -> JSRegEx a' x
+    JSArrayLiteral _ x _ -> JSArrayLiteral a' (map (rjsa a') x) a'
```
Comment (jeapostrophe) [inline hs/src/Reach/JSUtil.hs:354]: I think you should implement `rjsa` on lists

File: hs/src/Reach/JSUtil.hs [hunk 4]
```diff
@@ -268,19 +342,69 @@ instance HasJSAnnot JSExpression where
...
+    JSCallExpression a _ y _ -> JSCallExpression (rjsa a' a) a' (rjsaJSL a' y) a'
+    JSCallExpressionDot a _ y -> JSCallExpressionDot (rjsa a' a) a' (rjsa a' y)
+    JSCallExpressionSquare a _ y _ -> JSCallExpressionSquare (rjsa a' a) a' (rjsa a' y) a'
+    JSClassExpression _ x y _ r _ -> JSClassExpression a' (rjsa a' x) y a' r a'
+    JSCommaExpression a _ y -> JSCommaExpression (rjsa a' a) a' (rjsa a' y)
+    JSExpressionBinary a x y -> JSExpressionBinary (rjsa a' a) (rjsa a' x) (rjsa a' y)
+    JSExpressionParen _ x _ -> JSExpressionParen a' (rjsa a' x) a'
+    JSExpressionPostfix a x -> JSExpressionPostfix (rjsa a' a) (rjsa a' x)
+    JSExpressionTernary a _ y _ r -> JSExpressionTernary (rjsa a' a) a' (rjsa a' y) a' (rjsa a' r)
+    JSArrowExpression x _ y -> JSArrowExpression (rjsa a' x) a' $ jsStmtToConciseBody (rjsa a' $ jsArrowBodyToStmt y)
```
Comment (jeapostrophe) [inline hs/src/Reach/JSUtil.hs:366]: I think it is a mistake to do these conversions

File: hs/src/Reach/JSUtil.hs [hunk 5]
```diff
@@ -320,6 +444,90 @@ instance HasJSAnnot JSStatement where
...
+    JSExpressionStatement e x -> JSExpressionStatement (rjsa a' e) (rjsa a' x)
+    JSAssignStatement e x y z -> JSAssignStatement (rjsa a' e) (rjsa a' x) (rjsa a' y) (rjsa a' z)
+    JSMethodCall r _ x _ z -> JSMethodCall (rjsa a' r) a' (rjsaJSL a' x) a' (rjsa a' z)
+    JSReturn _ x y -> JSReturn a' (fmap (rjsa a') x) (rjsa a' y)
+    JSSwitch _ _ y _ _ s _ u -> JSSwitch a' a' (rjsa a' y) a' a' s a' (rjsa a' u)
+    JSThrow _ x y -> JSThrow a' (rjsa a' x) (rjsa a' y)
+    JSTry _ x y z -> JSTry a' (rjsa a' x) (map (rjsa a') y) (rjsa a' z)
+    JSVariable _ x y -> JSVariable a' (rjsaJSL a' x) (rjsa a' y)
+    JSWhile _ _ y _ r -> JSWhile a' a' (rjsa a' y) a' (rjsa a' r)
+    JSWith _ _ y _ r s -> JSWith a' a' (rjsa a' y) a' (rjsa a' r) (rjsa a' s)
```
Comment (jeapostrophe) [inline hs/src/Reach/JSUtil.hs:484]: I think you should do something like `where r = rjsa a'` and then you'll be able to write stuff like `JSFunction a' (r x) a' (r z) a' (r s) (r t)`

File: hs/t/n/API_OutBeforeIn.txt [hunk 6]
```diff
@@ -5,9 +5,9 @@ Verifying for generic connector
 Checked 4 theorems; No failures!
 reachc: error[RAPI0001]: API "Writer_f" calls interact.out() before interact.in()
 
-  ./API_OutBeforeIn.rsh:1:21:application
+  ./API_OutBeforeIn.rsh:12:19:application
 
-  1| 'reach 0.1';
+  12|   U.f.interact.out([], null);
```
Comment (jeapostrophe) [inline hs/t/n/API_OutBeforeIn.txt:10]: So awesome

File: hs/t/n/lifted_interact_non_null_fun.txt [hunk 7]
```diff
@@ -1,8 +1,8 @@
 reachc: error[RE0003]: Invalid block result type. Expected Null, got UInt
 
-  ./lifted_interact_non_null_fun.rsh:1:21:application
+  ./lifted_interact_non_null_fun.rsh:8:19:application
 
-  1| 'reach 0.1';
+  8|       A.interact.f();
```
Comment (jeapostrophe) [inline hs/t/n/lifted_interact_non_null_fun.txt:5]: Beautiful

**Review by jeapostrophe — CHANGES_REQUESTED** [review]

**Review by jeapostrophe — CHANGES_REQUESTED** [review]

File: hs/src/Reach/JSUtil.hs [hunk 8]
```diff
@@ -181,8 +181,11 @@ class HasJSAnnot a where
 class RepJSAnnot a where
   rjsa :: JSAnnot -> a -> a
 
-rjsaJSL :: (HasJSAnnot a, RepJSAnnot a) => JSAnnot -> JSCommaList a -> JSCommaList a
-rjsaJSL a' y =  toJSCL $ map (rjsa a') $ jscl_flatten y
+rjsaJSL :: RepJSAnnot a => JSAnnot -> JSCommaList a -> JSCommaList a
```
[the remark this answers is not in the document]
Comment (chikeabuah) [reply hs/src/Reach/JSUtil.hs:184]: yes

File: hs/src/Reach/JSUtil.hs [hunk 9]
```diff
@@ -448,9 +453,9 @@ instance RepJSAnnot JSStatement where
   rjsa a' = \case
     JSStatementBlock _ x _ z -> JSStatementBlock a' (fmap (rjsa a') x) a' (rjsa a' z)
     JSBreak _ x y -> JSBreak a' (rjsa a' x) (rjsa a' y)
-    JSLet _ x y -> JSLet a' (toJSCL (map (rjsa a') $ jscl_flatten x)) (rjsa a' y)
+    JSLet _ x y -> JSLet a' (toJSCL (rjsa a' $ jscl_flatten x)) (rjsa a' y)
```
[the remark this answers is not in the document]
Comment (chikeabuah) [reply hs/src/Reach/JSUtil.hs:456]: understood. the remaining `fmap`s are `Maybe`s

[the remark this answers is not in the document]
Comment (jeapostrophe) [reply hs/src/Reach/JSUtil.hs:456]: Implement `instance RepJSAnnot a => RepJSAnnot (Maybe a)`

[the remark this answers is not in the document]
Comment (chikeabuah) [reply hs/src/Reach/JSUtil.hs:456]: ah, will do

File: hs/t/n/API_OutBeforeIn.txt (hunk 6 above)
Comment (chikeabuah) [reply hs/t/n/API_OutBeforeIn.txt:10]: ^_^

File: hs/src/Reach/JSUtil.hs (hunk 4 above)
Comment (chikeabuah) [reply hs/src/Reach/JSUtil.hs:371]: okay

**Review by jeapostrophe — APPROVED** [review]

**Pull request merged**